Fix trailing parens tripping up skipToLineEnd

solarized-fox opened the pull request:
## Description
fixes #13017

it seems that parens are not included in the start-end range a node has for referencing the source text. this can cause skipToLineEnd to compute incorrectly because it isn't expecting to encounter any trailing ")"s.  adding ")" to the list of characters it skips fixes this

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

<!-- Please DO NOT remove the playground link -->

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

fisker commented:
Honestly, I didn't expect you to change this part ... 

It's also a public API.
If we accept this change, we'll need [this changelog](https://github.com/prettier/prettier/blob/main/changelog_unreleased/api/14317.md#update-prettierutil-14317-14320-by-fisker) that `skipToLineEnd` and `isNextLineEmpty` skips `)`.

But before that, we need to think about what will be affected.

fisker commented:
This case still broken


**Prettier pr-14736**
[Playground link](https://deploy-preview-14736--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBtAOlABFgFAQwF4AjTHMrASgBoLjqsB6RzAXQG4RqQIAHGAJbQAzslD4AThIgB3AAqSEolPgA2M-AE9R3YhPxgA1nBgBlfAFs4AGQFQ4yAGZrhcXfqMnTvA3YDmyDASAK5uIK4WAoEhYcL+qnAAisEQ8E4uYQBWwgAepvFJKWlIzqqu3ACORXBy0rzKIPjCALT2cAAmHVwgQfgCqv4AwhAWFvjIjaqq3XFQfgkAgjBBAsTB8HJwErb26WVhABYwFqoA6gcC8MI+YHCmSpcCAG6XmhNgwjogT6EAklCdWCmMASAT8BYA0wwTQJPblEC8aSuU76XgTRFwVwSJ4Obh2LEwWr4PxjOFhHwSLETYj4YhwabcRF2GCnATtGAHZAADgADNwJHAqgIBUSSeMShluDBaaz2ZykAAmbjBVwAFVpylK8LgFjp7U67Ws+DmwWJcAAYhAJGNlv4Jvh1hAQABfZ1AA)
<!-- prettier-ignore -->
```sh
--parser babel
```

**Input:**
<!-- prettier-ignore -->
```jsx
[
  (a=b
  
  ),
  b, //
];
```

**Output:**
<!-- prettier-ignore -->
```jsx
[
  (a = b),

  b, //
];

```

solarized-fox commented:
> Honestly, I didn't expect you to change this part ...

I tried a few other approaches, but they caused tests to fail. if you have any ideas I'm open to suggestions!

> But before that, we need to think about what will be affected.

as for this, my understanding of skipToLineEnd is that it supposes that the only non-comment characters between the last node and EOL are in the character class of "\t ,; ", which would mean that since ")" can appear in there, this is not actually a behavior change so much as a bug fix. that's why I didn't make a changelog, but I can add one if you believe it's important

as for:

> This case is not format correctly
> 
> **Input:**
> 
> ```js
> [
>   (a=b
>   
>   ),
>   b, //
> ];
> ```
> 
> **Output:**
> 
> ```js
> [
>   (a = b),
> 
>   b, //
> ];
> ```

I'm not sure the information required to fix this is present in the AST? the existing code uses the next line being whitespace as a sort of hacky heuristic here, but as you've demonstrated that is inadequate. any robust solution would require a fair bit of code to determine what is and isn't part of a multiline expression. perhaps this is better off raised as a followup issue, as it doesn't seem idempotency is violated in this example and fixing this beyond all edge cases might be a nontrivial effort requiring discussion?

solarized-fox commented:
I decided to force-push a different version of this change. it solves the edge case mentioned by @fisker and doesn't change any public APIs. if anyone knows a shorter way to do this I'm all ears but otherwise I think this is an improvement over the last solution

solarized-fox commented:
I noticed there was a trivial merge conflict (printArrayItems was renamed to printArrayElements while I was working on it) so I fixed that up and force-pushed the change. it should be automatically mergeable now

fisker commented on src/language-js/print/array.js:
```diff
@@ -171,6 +172,36 @@ function isConciselyPrintedArray(node, options) {
+      } else {
+        const nextCandidate =
+          text[currentIdx] == ")" ? currentIdx + 1 : candidateIdx;
+        return skipParens(skipIfComment(currentIdx + 1), nextCandidate);
```
Can `skip` function do the job?

solarized-fox replied:
I think it can't replace `skipParens` (and I think I'll rename `skipParens` to `indexAfterFurthestParen` for clarity) however after looking into it `skipIfComment` can be implemented with `skip` calls which cuts down a fair bit of code, so I'll make that change

fisker commented on src/language-js/print/array.js:
```diff
@@ -171,6 +172,36 @@ function isConciselyPrintedArray(node, options) {
+  function isLineAfterElementEmpty(elementNode) {
+    const text = options.originalText;
+    const elementIdx = elements.findIndex((v) => v == elementNode);
+    const nextNode = elements.find((e, i) => i > elementIdx && e !== null);
```
path.siblings, path.next, path.index should be used

solarized-fox replied:
oh, you're right, that does make sense, making these changes now

solarized-fox commented:
while working editing the code, I realized instead of skipping the parens I can just skip directly to the comma. this is a lot less code overall

however, it did trigger one change in the tests where
```
[
  1
  ,
  
  2,
]

used to format as `[1, 2]` and now formats as `[1,\n2]`.  this seems like it should be the correct behavior though, because otherwise a newline gets eaten. I have added a changelog and updated tests accordingly

solarized-fox commented:
thinking about the change, I guess the root question is, for this input:
```
_ = [
  (a = b
  //comment
  ),

  2,
  3 //
]
```

which of these outputs is most appropriate?

`git checkout #13017`
```
_ = [
  (a = b),

  //comment
  2,
  3, //
];
```

`git checkout #13017-2`
```
_ = [
  (a = b),
  //comment
  2,
  3, //
];
```

fisker commented:
Expected output should be

```js
_ = [
  (a = b), //comment

  2,
  3, //
];
```

It's bug in comment attach. Should be done in a seperate PR.

solarized-fox commented:
I see. well, setting aside the comment then, for this input:
```
_ = [
  (a = b
  ),

  2,
  3 //
]
```

which of these is preferred?

`git checkout #13017`
```
_ = [
  (a = b),

  2,
  3, //
];
```
`git checkout 13017-2`
```
_ = [
  (a = b),
  2,
  3, //
];
```

fisker commented:
```js
_ = [
  (a = b),

  2,
  3, //
];
```


Because there is a blank line between comma after first element and second element.

solarized-fox commented:
alright then, I'm fairly confident that this (the branch this PR is attached to) is the correct change. solving the root issue (i.e. #13017) without changing any snapshots would require writing code that then needs to be thrown out later. if you find anything else in the PR that needs changing, let me know, but otherwise I think it's ready

solarized-fox commented:
oops, saw the lint failure, fixed it and ran lint locally too so it should be solved for good

fisker commented on src/language-js/print/array.js:
```diff
@@ -170,6 +171,18 @@ function isConciselyPrintedArray(node, options) {
+  const skipToComma = (currentIdx) =>
+    text[currentIdx] === ","
+      ? currentIdx
+      : skipToComma(skipComment(currentIdx + 1));
```
Will this work?

```js
const skipToComma = skip(/[^,]/);
```

fisker reviewed (approved):
LGTM, thanks!

solarized-fox commented:
ah, sorry for the sudden force push, I accidentally pushed the button to merge changes from main into this branch, I rolled that back. is there something else I need to do or will these changes be merged at some point in the future?

sosukesuzuki commented on src/language-js/print/array.js:
```diff
@@ -170,6 +171,18 @@ function isConciselyPrintedArray(node, options) {
   );
 }
 
+function isLineAfterElementEmpty(path, { originalText: text }) {
```
Can we use destructuring assignment here?

```suggestion
function isLineAfterElementEmpty({ node }, { originalText: text }) {
```

sosukesuzuki reviewed (approved):
LGTM! Please check one comment.

fisker commented on src/language-js/print/array.js:
```diff
@@ -9,17 +9,18 @@ import {
 } from "../../document/builders.js";
 import hasNewline from "../../utils/has-newline.js";
+import isNextLineEmptyAfterIndex from "../../utils/is-next-line-empty.js";
+import { skipInlineComment, skipTrailingComment } from "../../utils/public.js";
```
Missed this on previous review. Do not import from `public.js`, please import from the actual file directly.

solarized-fox commented:
I've implemented the suggested changes!

solarized-fox commented:
oops, forgot a semicolon I guess
